Reduce logs in security/pkg/nodeagent/sds/

lei-tang opened the pull request:
https://github.com/istio/istio/issues/13033

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13035#" title="Author self-approved">lei-tang</a>*
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **myidpt**

If they are not already assigned, you can assign the PR to them by writing `/assign @myidpt` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[security/OWNERS](https://github.com/istio/istio/blob/master/security/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["myidpt"]} -->

lei-tang commented:
/cc @myidpt

howardjohn commented:
This doesn't fix the issue, all of the error messages are still there. I guess it cuts it in half though

howardjohn commented:
What happens if `.Serve()` never succeeds? There is going to be millions of logs produced?

lei-tang commented:
Error messages are useful for debugging purpose.  If .Serve() never succeeds, the error message will be output consistently and consecutively (instead of drown out among other logs), which highlights the Serve() problem. If a user does not want to see verbose logs, the user may choose different log levels, or pipe logs through filters to filter out the logs that the user are not interested in.

istio-testing commented:
@lei-tang: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | ad51fba22d51b2e7135f05991b086e6a63c8717f | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10529) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
> Error messages are useful for debugging purpose. If .Serve() never succeeds, the error message will be output consistently and consecutively (instead of drown out among other logs), which highlights the Serve() problem. If a user does not want to see verbose logs, the user may choose different log levels, or pipe logs through filters to filter out the logs that the user are not interested in.

Getting 1 million of the same error message is not helpful for debugging purposes. If making it "not drown out among other logs" was a good idea, then should we have every error message log 1000 times so they don't get drowned out? These are logged at the error level, so changing the log level isn't really a good idea...

These are logged in passing unit tests. A **passing** unit test should not log 100k error messages....

howardjohn commented:
This logs 100k times in a single second, its not like it logs this once per second

istio-testing commented:
@lei-tang: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | ad51fba22d51b2e7135f05991b086e6a63c8717f | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10529) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | ad51fba22d51b2e7135f05991b086e6a63c8717f | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-integ-k8s-tests/6182) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

lei-tang commented:
/cc @quanjielin

wattli commented:
Agree with Howard's point, i think it makes sense to have exponential + threshold, i.e., if the server can never be brought up, we should shut down the instance to signal error message.

lei-tang commented:
@howardjohn "This logs 100k times in a single second", which log is outputed 100k times in a single second? If so, this log should be removed due to 1) outputed 100k times in a single second decreases the program performance. 2) makes it difficult to view other logs.

howardjohn commented:
Just run `go test ./security/pkg/nodeagent/sds/...` or see example here: https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/12872/istio-unit-tests/23157

lei-tang commented:
Added counters to limit the max times error messages can be outputed. Please take a look. If this counting mechanism is not enough, a more complicated backoff mechanism can be devised and added.

incfly commented on security/pkg/nodeagent/sds/server.go:
```diff
@@ -193,15 +202,23 @@ func (s *Server) initGatewaySdsService(options *Options) error {
 
 	go func() {
+		log.Info("Start SDS grpc server for ingress gateway proxy")
+		numErrorOutputTime1 := 0
```
Time1 and Time2 are not good name for var.

lei-tang replied:
Done: revised based on the comments.

incfly commented on security/pkg/nodeagent/sds/server.go:
```diff
@@ -193,15 +202,23 @@ func (s *Server) initGatewaySdsService(options *Options) error {
 	}
 
 	go func() {
+		log.Info("Start SDS grpc server for ingress gateway proxy")
```
Please change to expotential back off, which is really simple. Psedu code.

```go
waitTime := time.Second
for i := 0; i < 10; i++ {
  log.Errorf("Failed to start gRPC server for SDS, retry in %v", waitTime)
  time.Sleep(waitTime)
  waitTime *= 2
}
```

lei-tang replied:
Done: the PR has been revised accordingly.

incfly commented on security/pkg/nodeagent/sds/server.go:
```diff
@@ -165,15 +166,23 @@ func (s *Server) initWorkloadSdsService(options *Options) error { //nolint: unpa
-			log.Info("Start SDS grpc server")
 			if err = s.grpcWorkloadServer.Serve(s.grpcWorkloadListener); err != nil {
-				log.Errorf("SDS grpc server for workload proxies failed to start: %v", err)
+				if numErrorOutputTime1 < maxErrorMessageOutputTimes {
```
What's the typical reason retry would help? If it's host file mount readiness issue, sleep can help here.

lei-tang replied:
Sleeping mechanism is added.

istio-testing commented:
@lei-tang: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | ad51fba22d51b2e7135f05991b086e6a63c8717f | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-integ-k8s-tests/6182) | `/test istio-integ-k8s-tests`
prow/istio-pilot-multicluster-e2e.sh | 915fe56275302671c09722ed70fa7dc17181815d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10530) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@lei-tang: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 915fe56275302671c09722ed70fa7dc17181815d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10530) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 915fe56275302671c09722ed70fa7dc17181815d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-integ-k8s-tests/6185) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

lei-tang commented:
/cc @wattli

istio-testing commented:
@lei-tang: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-integ-k8s-tests.sh | 915fe56275302671c09722ed70fa7dc17181815d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-integ-k8s-tests/6185) | `/test istio-integ-k8s-tests`
prow/istio-pilot-multicluster-e2e.sh | 7745f09def437aef5fdda69cea04580db644ed1b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10533) | `/test istio-pilot-multicluster-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@lei-tang: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-multicluster-e2e.sh | 7745f09def437aef5fdda69cea04580db644ed1b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-pilot-multicluster-e2e/10533) | `/test istio-pilot-multicluster-e2e`
prow/istio-integ-k8s-tests.sh | 7745f09def437aef5fdda69cea04580db644ed1b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pr-logs/pull/istio_istio/13035/istio-integ-k8s-tests/6187) | `/test istio-integ-k8s-tests`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/13035#" title="Author self-approved">lei-tang</a>*, *<a href="https://github.com/istio/istio/pull/13035#pullrequestreview-222498982" title="LGTM">quanjielin</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[security/OWNERS](https://github.com/istio/istio/blob/master/security/OWNERS)~~ [quanjielin]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

incfly commented:
Thanks @lei-tang for the fix.

> with this change, we need a way to stop nodeagent when exit go func(){}, so that k8s could restart it.
otherwise nodeagent will hang there after hitting maxRetryTimes in real env. We could have a different PR for this.

@quanjielin  Not sure what you mean. In this current code, I expect when maxRetry hit, node agent will finish complete automatically. Or you want to explicit use another os.Exit(-1) non zero return code, which is necessary for k8s to restart a pod?

By the way, I don't see why we need to use a goroutine for `server.Serve()` though. It's invokved from default main goroutine, main.go, `sds.NewServer`. We can just let it hanging there using default goroutine. And return error if needed.

Thoughts?

lei-tang commented:
https://github.com/istio/istio/issues/13439
